text/3389-manifest-lint.md

```diff
@@ -0,0 +1,428 @@
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Since `[lints]` only affects the associated package, and not dependencies, it
+will not work with `future-incompat` lints that are meant to be applied to
+dependencies.  This may cause some user confusion.
+
+There has been some user/IDE confusion about running commands like `rustfmt`
+directly and expecting them to pick up configuration only associated with their
+higher-level cargo-plugins despite that configuration (like `package.edition`)
+being cargo-specific.  By baking the configured lint levels for rustc, rustdoc, and
+clippy directly into cargo, we will be seeing more of this.  A hope is that
+this will actually improve with this RFC.  Over time, tools will need to switch
+to the model of running `cargo` to get configuration in response to this RFC.
+As for users, if a tool's primary configuration is in `Cargo.toml`, that will
+provide a strong coupling with `cargo` in users minds as compared to using an
+external configuration file and overlooking the one or two fields read from
+`Cargo.toml`.
+
+As this focuses on lints, this leaves out first-party tools that need
+configuration but aren't linters, namely `rustfmt`, leading to an inconsistent
+experience if `clippy.toml` goes away in the future (if we act on the future
+possibility of supporting linter configuration)
+
+A concern brought up in
+[rust-lang/rust-clippy#1313](https://github.com/rust-lang/rust-clippy/issues/1313)
+was that this will pass lints unconditionally to the underlying tool, leading
+to "undefined lint" warnings when used on earlier versions, requiring that
+warning to also be suppressed, reducing its value.  However, in the "Future
+possibilities" section, we mention direct support for tying lints to rust versions.
+
+This does not allow sharing lints across workspaces.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+This could be left to `clippy.toml` but that leaves `rustc` without a solution.
+
+`[lints]` could be `[package.lints]`, tying it to the package unlike `[patch]`
+and other fields that are more workspace related.  Instead, we used
+`[dependencies]` as our model.
+
+`[lints]` could be `[lint]` but we decided to follow the precedence of `[dependencies]`.
+
+Instead of `<tool>.<lint>`, we could use `<tool>::<lint>` (e.g.
+`"clipp::enum_glob_use"` instead of `clippy.enum_glob_use`), like in the
+diagnostic messages.  This would make it easier to copy/paste lint names.
+However, with the schema being `<lint> = <level>`, this would require quoting
+the keys rather than leaving them as bare words.  This would also cause
+problems for tool-level configuration.  The first is that the lints wouldn't be
+easily grouped with their config.  The second is if we use `<lint> = <level>`,
+we'd be mixing tool names in with lints, making it easier to conflict and
+harder to collect all lints to forward to a tool.
```

Comment:
It's not necessary to collect the lints per-tool, tools automatically ignore all lints that aren't in their namespace

```rust
> cargo rustdoc -- -W clippy::pedantic
 Documenting cbor-diag v0.1.11 (/home/nemo157/sources/cbor-diag-rs)
    Finished dev [unoptimized + debuginfo] target(s) in 0.43s
> cargo clippy -- -W rustdoc::foobaz
    Checking cbor-diag v0.1.11 (/home/nemo157/sources/cbor-diag-rs)
    Finished dev [unoptimized + debuginfo] target(s) in 0.76s
```

The potential exception to this could be `cargo::` lints, since that isn't a registered tool namespace, and doesn't really make sense to be as the lints won't be configured in source code.

Reply:
This is talking about the data being
```toml
[lints]
"clippy::pedanic" = "warn"

[lints.clippy]
rust-version = "1.60.0" 
```
which maps to
```json
{
  "lints": {
    "clippy": {
      "rust-version": "1.60.0",
    }
    "clippy](clippy::pedanic": "warn",
  }
}
```

Now, we need to collect the lints from this to pass to the tools but which keys are for lints?  We can try to use the shape of the data but that seems brittle especially if we support lint configuration.

Reply:
Ah, this is assuming the future extension of lint config. I'm not sure that it makes sense to have specifically lint config, a more general tool config could be better, that would allow configuring things other than lints, and could also be used for configuring clippy's lints:

```toml
[tool.rustdoc]
show-type-layout = true
```

Having the config under a separate section would avoid this conflict.

Reply:
c3f932cc tries to remove some of the confusion around this.